runtime/src/main/cpp/Memory.cpp

```diff
@@ -2262,56 +2287,67 @@ void freezeAcyclic(ContainerHeader* rootContainer, ContainerHeaderSet* newlyFroz
-    current->unMark();
-    reversedEdges.emplace(current, KStdVector<ContainerHeader*>(0));
-    traverseContainerReferredObjects(current, [current, &queue, &reversedEdges](ObjHeader* obj) {
+    ContainerHeader* currentContainer = current->container();
+    currentContainer->unMark();
+    reversedEdges.emplace(currentContainer, KStdVector<ContainerHeader*>(0));
+    traverseContainerReferredObjects(currentContainer, [current, currentContainer, &queue, &reversedEdges](ObjHeader* obj) {
           ContainerHeader* objContainer = obj->container();
           if (canFreeze(objContainer)) {
             if (objContainer->marked())
-              queue.push_back(objContainer);
-            reversedEdges.emplace(objContainer, KStdVector<ContainerHeader*>(0)).first->second.push_back(current);
+              queue.push_back(obj);
+            // We ignore FreezableAtomicsReference during condensation, to avoid KT-33824.
+            if (!isFreezableAtomic(current) && !isFreezableAtomic(obj))
+              reversedEdges.emplace(objContainer, KStdVector<ContainerHeader*>(0)).
+                first->second.push_back(currentContainer);
           }
       });
-    }
+   }
 
-    KStdVector<KStdVector<ContainerHeader*>> components;
-    MEMORY_LOG("Condensation:\n");
-    // Enumerate in the topological order.
-    for (auto it = order.rbegin(); it != order.rend(); ++it) {
-      auto* container = *it;
-      if (container->marked()) continue;
-      KStdVector<ContainerHeader*> component;
-      traverseStronglyConnectedComponent(container, &reversedEdges, &component);
-      MEMORY_LOG("SCC:\n");
+   KStdVector<KStdVector<ContainerHeader*>> components;
+   MEMORY_LOG("Condensation:\n");
+   // Enumerate in the topological order.
+   for (auto it = order.rbegin(); it != order.rend(); ++it) {
+     auto* container = *it;
+     if (container->marked()) continue;
+     KStdVector<ContainerHeader*> component;
+     traverseStronglyConnectedComponent(container, &reversedEdges, &component);
+     MEMORY_LOG("SCC:\n");
   #if TRACE_MEMORY
-      for (auto c: component)
-        konan::consolePrintf("    %p\n", c);
+     for (auto c: component)
+       konan::consolePrintf("    %p\n", c);
   #endif
-      components.push_back(std::move(component));
-    }
+     components.push_back(std::move(component));
+   }
 
-    // Enumerate strongly connected components in reversed topological order.
+  // Enumerate strongly connected components in reversed topological order.
   for (auto it = components.rbegin(); it != components.rend(); ++it) {
     auto& component = *it;
     int internalRefsCount = 0;
     int totalCount = 0;
     for (auto* container : component) {
+      RuntimeAssert(!isAggregatingFrozenContainer(container), "Must not be called on such containers");
       totalCount += container->refCount();
+#if TRACE_MEMORY
+     if (isFreezableAtomic(container)) {
+       RuntimeAssert(component.size() == 1, "Must be trivial condensation");
+       continue;
```

Comment:
Isn't this required if `TRACE_MEMORY` is not true?

Reply:
It is just an excessive check, not sure if it makes sense in non-tracing case.

Reply:
I guess the tests are failing due to this.